### ISSUE TITLE
Untranslatable strings

### ISSUE THREAD

**Issue by Matthaiks** [mumble-voip/mumble#4359]
**Describe the bug**
Some strings are in English only.

**Steps to Reproduce**
1. Switch to Dutch or Polish (100% translated locales according to Transifex)
2. Go to Configure -> Settings... -> Audio Output

"Attenuate other users while talking as Priority Speaker" is in English despite being translated.

3. Go to Server -> Connect... -> Add New... 

"Ignore" is in English, this string seems to be missing from the language files.

**Expected behavior**
Fix hard-coded "Attenuate other users while talking as Priority Speaker" and make translating "Ignore" possible.

**Desktop (please complete the following information):**
 - OS: Windows 10 x64 1809
 - Version: Mumble 1.3.2

**Comment by Krzmbrzl** [comment]
The translations are for the current master branch.

The 1.3.x series has been released in September of last year whereas the mentioned `Attenuate other users while talking as Priority Speaker` string has been translated 7 months ago.

And we don't backport translations as this would require sorting out which strings were already present in the 1.3.x series and separating those from the ones that are new in 1.4.0

TL;DR: This strings aren't hardcoded, they have just not been translated in time for 1.3.x

> "Ignore" is in English, this string seems to be missing from the language files.

This one is indeed hardcoded because someone (I assume accidentally) un-checked the respective checkbox in Qt Designer. I'll quickly fix this. Thanks for reporting :+1:

**Comment by Matthaiks** [comment]
"Attenuate other users while talking as Priority Speaker" is translated in 1.3.0, though.

**Comment by Krzmbrzl** [comment]
@Matthaiks in Dutch?

**Comment by Krzmbrzl** [comment]
Ah wait a sec. It is translated in 1.3.0 but it is not translated in 1.3.2? :eyes:

**Comment by Matthaiks** [comment]
Yeah, 1.3.0 has it translated in multiple languages.

**Comment by Krzmbrzl** [comment]
Is it translated **in Dutch** in 1.3.0 but not 1.3.2? :thinking:

**Comment by Matthaiks** [comment]
Yes.

**Comment by Krzmbrzl** [comment]
Indeed. Then the history for this entry on Transifex doesn't apply for some reason...

The translation is also still present in the code-state of 1.3.2 (in the file).

I don't really understand why it wouldn't show up in the UI :thinking:

**Comment by Matthaiks** [comment]
Mumble-1.4.0\~2020-12-24\~gdb4b6be\~snapshot.winx64.msi

"Attenuate other users while talking as Priority Speaker" and "Ignore" are translated now.

"Show nicknames only" and "Hide the username for each user if they have a nickname." remain in English. Those strings are translated if there's the .qm file in the app folder, though.

Bigger issue: translations of Qt seem to be missing, so lots of buttons and context menu items are in English.

**Comment by Krzmbrzl** [comment]
> Bigger issue: translations of Qt seem to be missing, so lots of buttons and context menu items are in English.

Afaik we don't touch these. But maybe the packaging process prevents them from being included :thinking: 
@davidebeatrici @ZeroAbility ideas?

> "Show nicknames only" and "Hide the username for each user if they have a nickname." remain in English. Those strings are translated if there's the .qm file in the app folder, though.

I am pretty sure the nicknames entry is not translated yet (on Transifex and thus in the respective .ts file). What .qm file are you referring to? :eyes:

**Comment by Matthaiks** [comment]
![TransifexMumblePL](https://user-images.githubusercontent.com/3577122/103226779-218a4d80-492d-11eb-8bd4-34fea7cece8d.png)
I've compiled the .qm file from the Transifex's .ts file

**Comment by Krzmbrzl** [comment]
Ah okay :+1:

**Comment by Matthaiks** [comment]
Audacious had a problem with the Qt translations. Maybe this would be useful: https://redmine.audacious-media-player.org/issues/569

**Comment by Krzmbrzl** [comment]
We are actually doing (pretty much) the same thing as in the patch in the linked audacious report. That should be the case since the latest development snapshot (the third one).
I assume you have checked that these Strings are still not translated in that version?

**Comment by Matthaiks** [comment]
`Mumble-1.4.0~2021-03-20~gf8dc8b0~snapshot.x64`

![Mumble (2021 03 25, 12-43-08)](https://user-images.githubusercontent.com/3577122/112467829-d040fd00-8d67-11eb-81bd-b36a51eb9897.png)

As you can see, still not translated.

**Comment by Krzmbrzl** [comment]
@davidebeatrici what was the reasoning for not bundling Qt translations in Mumble?

**Comment by Kissaki** [comment]
This works fine in 1.3

![image](https://user-images.githubusercontent.com/93181/112521162-7c9cd680-8d9c-11eb-9a56-5b24d4d44f67.png)

So I suspect this is a CMake transition regression?

**Comment by Krzmbrzl** [comment]
Probably. But the question is what has changed in terms of translations? :thinking: 
@davidebeatrici any ideas?

**Comment by Kissaki** [comment]
Or maybe translation path loading changes? Is this only the last snapshot?

**Comment by Krzmbrzl** [comment]
No it existed before already

**Comment by Krzmbrzl** [comment]
There is also #4895 which seems to be the same issue that remains here :thinking:

**Comment by davidebeatrici** [comment]
Let's bundle Qt translations again.

**Comment by davidebeatrici** [comment]
We add the files from https://github.com/mumble-voip/mumble/tree/master/src/mumble/qttranslations to: https://github.com/mumble-voip/mumble/blob/06f3b3227831abe0cdc023f6090e212363295e5d/src/mumble/CMakeLists.txt#L713-L754

**Comment by Krzmbrzl** [comment]
But these seem to only be a small portion of the possible Qt translations. For instance German is not in there at all.
So why do the translations for these locales require special care? And most importantly: Would that fix the issue for the other locales as well? I would expect not :thinking:

**Comment by Krzmbrzl** [comment]
I am having the impression that in order for the standard Qt Strings to be translated, the user has to have these Qt translations installed in the default directory on their machine. If they don't have that, they won't have translations for these Strings.

A possible explanation for this that I can come up with is that the old 1.3.x static environment had these translations bundled somehow whereas the vcpkg build of Qt does not bundle them and thus we lost translations going from 1.3.x to 1.4.x :thinking:

**Comment by davidebeatrici** [comment]
> But these seem to only be a small portion of the possible Qt translations. For instance German is not in there at all.
> So why do the translations for these locales require special care? And most importantly: Would that fix the issue for the other locales as well? I would expect not :thinking:

We only provide Qt translations that are either missing upstream or of poor quality.

> I am having the impression that in order for the standard Qt Strings to be translated, the user has to have these Qt translations installed in the default directory on their machine. If they don't have that, they won't have translations for these Strings.

Correct.

> A possible explanation for this that I can come up with is that the old 1.3.x static environment had these translations bundled somehow whereas the vcpkg build of Qt does not bundle them and thus we lost translations going from 1.3.x to 1.4.x :thinking: 

https://wiki.qt.io/How_to_create_a_multi_language_application

https://github.com/mumble-voip/mumble/blob/f5bf5f17f3c207bb18ecd2dfc9ccdfa3136b6c7f/INSTALL#L148-L150

https://github.com/mumble-voip/mumble/blob/f5bf5f17f3c207bb18ecd2dfc9ccdfa3136b6c7f/src/mumble/mumble.pro#L692-L728

**Comment by Krzmbrzl** [comment]
Aha. So essentially I assume the cmake version  currently does not contain an equivalent of the code in the kast link you provided.
I guess we shoukd change that? :thinking:

**Comment by davidebeatrici** [comment]
Indeed.

**Comment by Krzmbrzl** [comment]
Bundling the translations is one thing. I just checked and on my system the `qtbase_de.qm` translation is loaded from the standard location but these buttons are still all in English.
This seems to indicate that the problem lies somewhere else.

Note that I tried this with loading `qtbase_de.qm` as well as with `qt_de.qm`. I also verified that these files do exist. Could someone check this on their machine (with a different locale)?
In order to check which Qt translation is used, add printout statements to `src/mumble/Translations.cpp`.

**Comment by deluxghost** [comment]
🤔 Will this issue be resolved before the 1.4.0 release?

**Comment by Krzmbrzl** [comment]
Only if we find a solurion until then :thinking:

**Comment by Krzmbrzl** [comment]
To get some movement in here I have just checked the old qmake code for bundling the overwrite translations. Including that should definitely be feasible.

In the grand scheme of things though it seems that the translation coverage of Qt itself is really, really poor: https://www.transifex.com/qtproject/qttranslations/
Thus I think that this is an issue that we can't circumvent unless we explicitly set text for standard buttons that we could then translate ourselves...

Overall I have to wonder though whether it really makes sense to overwrite Qt translations in our project instead of having potential translators fixing the issues upstream... I guess the issue is that Qt will not (regularly) update its builds with new translations? @davidebeatrici do you know something about this?

**Comment by davidebeatrici** [comment]
On Linux Qt translations are regularly updated because provided through the package manager and loaded dynamically by applications.

On Windows, instead, we used to bundle them before switching to CMake.

**Comment by Matthaiks** [comment]
SMPlayer (dev version for Windows) is now shipping with Qt 5.15.2 (also present in Mumble's latest snapshot) but its translations do work.

**Comment by davidebeatrici** [comment]
You're right, they provide Qt translations in the package:

![](https://user-images.githubusercontent.com/5897523/115765382-faa5ca80-a3a6-11eb-8785-a42457ca1fb4.png)

**Comment by Krzmbrzl** [comment]
> On Windows, instead, we used to bundle them before switching to CMake.

Did we bundle them all? 'cause in the repo we only had some translations to overwrite Qt ones...
Do you happen to have a link to the place in the qmake files where we are bundling the main Qt translations? :eyes:

**Comment by davidebeatrici** [comment]
See https://github.com/mumble-voip/mumble/issues/4359#issuecomment-812084054.

All translations were taken from the build environment, but we used to override some that were either incomplete or of poor quality.

**Comment by Krzmbrzl** [comment]
> See #4359 (comment).

No that is only the overwrite part. This does not bundle the main Qt translations.

> All translations were taken from the build environment,

Where? I can't find the part that grabs the translations from Qt's translation dir and bundles them into Mumble. :eyes:

**Comment by davidebeatrici** [comment]
311f6f7e86fad11b13ccfaa82d663c369ace16cb

**Comment by Krzmbrzl** [comment]
Ah yes indeed. I always overlooked that the script gets `QT_INSTALL_TRANSLATIONS` as a first parameter :facepalm: 
Thanks for the link!

**Comment by davidebeatrici** [comment]
No problem. The script is indeed a bit confusing.

**Comment by Krzmbrzl** [comment]
Okay the issue should get fixed with #4944. Could someone verify that this is indeed working as expected (@deluxghost maybe)?

**Comment by Matthaiks** [comment]
BTW, translating on Weblate is locked:
> Auto-merging src/mumble/mumble_zh_CN.ts
Auto-merging src/mumble/mumble_sv.ts
Auto-merging src/mumble/mumble_ru.ts
Auto-merging src/mumble/mumble_pl.ts
Auto-merging src/mumble/mumble_nl.ts
CONFLICT (content): Merge conflict in src/mumble/mumble_nl.ts
Auto-merging src/mumble/mumble_fi.ts
CONFLICT (content): Merge conflict in src/mumble/mumble_fi.ts
Removing src/mumble/Plugins.h
Removing src/mumble/Plugins.cpp
Removing .github/workflows/checks.yml
Automatic merge failed; fix conflicts and then commit the result.
 (1)

**Comment by deluxghost** [comment]
> Okay the issue should get fixed with #4944. Could someone verify that this is indeed working as expected (@deluxghost maybe)?

If there is a downloadable CI artifact to test, I will

Just too busy to build it myself recently

**Comment by Krzmbrzl** [comment]
I will give you the link once it has built successfully - we need to recompile our build env in order for the new changes to work xD

**Comment by Krzmbrzl** [comment]
@deluxghost after a lot of fiddling the build seems to be working now. Please let me know whether this does what it is intended to do :)

**Comment by Matthaiks** [comment]
Fixed!

![6347347347](https://user-images.githubusercontent.com/3577122/117585315-108aed80-b112-11eb-8d10-60fc71fb867c.png)



Yet another issue (also present in older 1.4.0 builds). Translations don't work for "Mixed echo cancellation (speex)" and "Multichannel echo cancellation (speex)":

![754325123623u](https://user-images.githubusercontent.com/3577122/117585430-ade62180-b112-11eb-9e4b-690166257b1c.png)

**Comment by deluxghost** [comment]
no qt translation for `zh-CN`, is it bundled?

**Comment by Krzmbrzl** [comment]
> Yet another issue (also present in older 1.4.0 builds). Translations don't work for "Mixed echo cancellation (speex)" and "Multichannel echo cancellation (speex)":

It could easily be that these Strings are not translated yet. You can head over to https://hosted.weblate.org/projects/mumble/ and translate them though.


@deluxghost iirc with `zh-CN` there was the issue that the Qt translation of it was really poor and we have to use our own version of it, right? Them not being shown then probably means that the overwriting does not work yet :thinking: 
I theory everything should be bundled though. I'll have another look.

**Comment by deluxghost** [comment]
> iirc with `zh-CN` there was the issue that the Qt translation of it was really poor and we have to use our own version of it, right? Them not being shown then probably means that the overwriting does not work yet :thinking: 

yes it was. although I will try to fix some translation on qt side (harder and takes more time), we still need this overwrite workaround now

**Comment by Matthaiks** [comment]
> It could easily be that these Strings are not translated yet. You can head over to https://hosted.weblate.org/projects/mumble/ and translate them though.

These strings were translated a long time ago.

**Comment by Krzmbrzl** [comment]
> These strings were translated a long time ago.

Hm that's odd.

@Kissaki do you have an idea what the issue could be? The translation is clearly present:
https://github.com/mumble-voip/mumble/blob/8e89110025eee14ddb81e0c2acaebd6b4339dbb1/src/mumble/mumble_nl.ts#L7500-L7503

And it should be used as of
https://github.com/mumble-voip/mumble/blob/baeddddec4c301389a088a9f7059a4b5d28c7dd1/src/mumble/EchoCancelOption.h#L31

but for some reason it isn't. I also verified that the translation is in the `QObject` context :thinking:

**Comment by Kissaki** [comment]
Given that this is a static const array, maybe this is an initialization order issue? Is it initialized without the correct language having been set?

_(I have not checked the code context here or whether we do that elsewhere where it works.)_

**Comment by Krzmbrzl** [comment]
Ah yes indeed that sounds very plausible :+1:

**Comment by Gert-dev** [comment]
I found another couple of strings that appear to be untranslatable - they are not translated in Dutch nor in German, and I can't seem to find them in Weblate either. Perhaps these will also be fixed by upcoming pull requests, but I thought I'd mention them here anyway, in case they aren't or so they can be checked:

In the settings window:
* `Restore Defaults`
* `Apply`
* `Cancel`
* `Ok`
* `Reset`
* (Strangely enough the other button to restore everything to default values _is_ translated.)

In the audio wizard:

* `Back`
* `Next`
* `Cancel`
* `Finish`

[Inspecting the code for the settings window](https://github.com/mumble-voip/mumble/blob/baeddddec4c301389a088a9f7059a4b5d28c7dd1/src/mumble/ConfigDialog.cpp#L70), I see that the button tooltips are translated, and that the button text of just the "Defaults (All)" button is translated, which seems to match what my findings.

**Comment by Krzmbrzl** [comment]
@Gert-dev these seem to all be Qt standard translations. They should get translated with bundling the Qt translations into the client binary as done in #4944.

Could you download the latest artifact to check if these Strings are translated with it?

@deluxghost I think I have managed to bundle the overwrite translations from our repo now. Could you also check the latest artifact to verify that it is working?

Latest artifacts: https://dev.azure.com/Mumble-VoIP/Mumble/_build/results?buildId=4550&view=artifacts&pathAsName=false&type=publishedArtifacts

**Comment by deluxghost** [comment]
@Krzmbrzl By a quick check I would say it's not working

I'm also looking at the QT part. I wonder:

- What's the correct way to translate it? Transifex?
- Which part does mumble actually use so I can work on them first.
   - only controls? which one? any others?
- Does mumble use the newest qt translation file, or only that shipped with qt release?

If this doesn't take a long time, maybe we can forget the overwrite thing?

**Comment by Krzmbrzl** [comment]
> By a quick check I would say it's not working (file SHA1 2acdbc1d4f15bb379b916d0087b632dee72b07da)

Meh :/

> What's the correct way to translate it? Transifex?

I guess? :shrug: 

> Which part does mumble actually use so I can work on them first. 

I think in general we mostly use Qt's standard buttons that require translation (plus their tooltips)

> Does mumble use the newest qt translation file, or only that shipped with qt release?

We use what is shipped with the vcpkg Qt package and thus I think we are using what has been released. I am not even sure though if there will be upcoming releases of Qt 5 though, given that Qt 6 is out now :thinking:

**Comment by Krzmbrzl** [comment]
@deluxghost please try again with https://dev.azure.com/Mumble-VoIP/Mumble/_build/results?buildId=4565&view=artifacts&pathAsName=false&type=publishedArtifacts
According to the build logs it should definitely have it bundled this time. If that does not work, I don't know what to do :shrug:

**Comment by deluxghost** [comment]
@Krzmbrzl cheers, it works! i will still help with qt part later tho

something unrelated:

![image](https://user-images.githubusercontent.com/5835759/118127196-f875df80-b42b-11eb-834f-b92a46798d00.png)

**Comment by Krzmbrzl** [comment]
Perfect!

That is probably another Qt font issue. Afaik there's nothing we can do about it :shrug:

**Comment by Gert-dev** [comment]
@Krzmbrzl I've built the latest master from 9e8a197f5cfc6689d1f4e5f5599087f9802b4ec4 (which seems to include this PR), but unfortunately these strings still don't appear to be translated.

It's possible that they still need to be translated on Weblate, though, but no untranslated strings show up, there. [Searching explicitly for e.g. "Restore Defaults"](https://hosted.weblate.org/translate/mumble/mumble-client/nl/?offset=1&q=%22Restore+Defaults%22&sort_by=-priority%2Cposition&checksum=) also doesn't show these strings. Perhaps they still have to be picked up by Weblate after this change?

**Comment by Krzmbrzl** [comment]
No asI said these appear to be Qt Strings that are translated in the Qt project and not in Mumble. We only bundle what Qt has so maybe these strings aren't translated in your language in Qt? :shrug:
@deluxghost are these Strings untranslated for you as well (with the nost recent changes active)? If so there might indeed be an issue on our end :thinking:

**Comment by deluxghost** [comment]
> are these Strings untranslated for you as well (with the nost recent changes active)? If so there might indeed be an issue on our end

@Krzmbrzl these strings are exactly what we were talking about in the issue, i.e. the QT translation. As I said, with the latest commit, it's fixed for zh-CN by force overwriting the file.

@Gert-dev what's your language? if you'd like to fix that, you can help QT project translate these strings like what I WILL do, eventually mumble will have these strings. you don't need to translate all strings for QT, only that mumble has used. Another way is bundle a patched QT TS file in mumble, like what I DID, but this might  be considered as a temporary or dirty workaround.

BTW, by helping QT project, not only mumble will get these translation, but other projects will get too.

**Comment by Krzmbrzl** [comment]
@Gert-dev in that case you could pull the Qt translation for your language from Transifex (or create a new language from their qs files), translate the necesaary Strings using e.g. Qt Linguist and then we could add that to our repo and bundle it in the same way that we do for the chinese translations from deluxghost. Only.l if you want of course :)

**Comment by Gert-dev** [comment]
@Krzmbrzl Are these the translations [located in the `qttranslations` folder in Mumble](https://github.com/mumble-voip/mumble/tree/master/src/mumble/qttranslations)? I see that there is already a `qt_nl.ts` in here, which seems to have those strings for `QDialogButtonBox` translated in Dutch. The file looks outdated and old, but doesn't seem to be applied - perhaps I'm looking at the wrong file?

**Comment by Krzmbrzl** [comment]
It is only used as a fallback translation as can be seen in the corresponding conf file:
https://github.com/mumble-voip/mumble/blob/8c99fe8119ce00fbcba55c69bde0a48383e7fa79/src/mumble/qttranslations/translations.conf#L19

That means that as long as there is an official version of the Dutch Qt translation available at the time of building Mumble, that will be used instead.

This behavior can be changed as was done for the Chinese translation:
https://github.com/mumble-voip/mumble/blob/8c99fe8119ce00fbcba55c69bde0a48383e7fa79/src/mumble/qttranslations/translations.conf#L21

**Comment by Gert-dev** [comment]
That sounds good. However, I tried changing that line and building Mumble myself, but these strings still don't appear translated. As a sanity check, I tried both pasting the contents of `qt_zh_CN.ts` into `qt_nl.ts` and actually changing my language to `zh_CN` in Mumble itself, but these strings also aren't translated in this language for me (obviously the rest of the interface is illegible to me :smile:), am I perhaps doing something wrong?

I tested this on both a Mumble Git build from the AUR (Arch Linux) and in a build I made myself locally, but both keep showing English strings for the dialog buttons, for both Dutch and Chinese.

**Comment by Krzmbrzl** [comment]
@Gert-dev you probably did neither set the `static` nor the `bundle-qt-translations` flag. I assume you are not building a static version, so you should invoke cmake as
```
cmake -Dbundle-qt-translations=ON ..
```

**Comment by Gert-dev** [comment]
I completely forgot about this, but got around to trying your latest comment, and that indeed was what I was missing. I created a PR with the updates.

Might I also suggest adding that option to the default build instructions or even enabling it by default (or inform why this is not the case)? I now also notice that a couple of the untranslated buttons were in fact already translated by these fallback settings, but they were never applied because the mumble-git package from the Arch Linux AUR doesn't build with this flag.

I can easily put a comment on the package to notify the package author of this problem, but it is very likely that the official Arch package and other distributions, once they update to 1.4.0 stable, will also omit this flag since it's not immediately clear it exists (and it also only appears to be useful to a couple of locales).

**Comment by Krzmbrzl** [comment]
Well in most cases it is desirable to use the latest Qt translations as installed on the respective system and I am not sure that these will still be consulted if these translations are bundled...

Besides: Fixing Qt's translations this way really is a crutch. Ideally translations should be contributed to Qt itself and then the Qt installation on your system should get updated as well.

Mentioning the existence of this option might be okay, but I am unsure as to where exactly to mention it. That being said though, this option is already listed in the [complete list of options](https://github.com/mumble-voip/mumble/blob/master/docs/dev/build-instructions/cmake_options.md#bundle-qt-translations).

**Comment by Gert-dev** [comment]
That does make sense, but this raises the question (I have no personal beef here, but am playing devil's advocate): why still maintain these translations at all, in that case? Since the option is not default, I'm assuming it's also not the default in the Windows build (if it is there, why not make it consistent and default everywhere). Not Arch Linux' official build nor the AUR package enable it. I'd be surprised if other Linux distributions were. In short: the only people benefiting from these translations are those enabling it manually at build time - like I'm doing here. This is a very small subset of people - wouldn't it be easier to just drop them from Mumble entirely?

At the same time, apparently some Qt translations are broken enough that they need fixing. The correct way to do this would indeed be to upstream the proper fixes - convoluted as the contributing process may be. Still, Chinese was configured as override because it's especially bad, IIUC - but what small subset of Chinese users are even getting these fixes due to the above problem, currently :smile:?

tl;dr: In other words: is there a point to having these translations in Mumble at all if they are not used by default? :smile:

**Comment by Krzmbrzl** [comment]
The mentioned option defaults to being ON as soon as the `static` option is used, which is always the case for Windows binaries.
On Linux the default is a dynamic build which is why it would have to be enabled manually (since it's a static feature).

That being said, I'd be first in line to throw this thing completely out the window and tell people to go fix Qt translations instead. Only because it was already this way when I joined the team and doesn't really hurt to keep around sm I even fiddling with this ^^
